src/identity_tree.rs

```diff
@@ -1,26 +1,247 @@
-use crate::timed_rw_lock::TimedRwLock;
+use std::{str::FromStr, sync::Arc};
+
 use semaphore::{
     merkle_tree::Hasher,
-    poseidon_tree::{PoseidonHash, PoseidonTree},
+    poseidon_tree::{PoseidonHash, PoseidonTree, Proof},
     Field,
 };
-use std::sync::Arc;
+use serde::Serialize;
+use thiserror::Error;
+use tokio::sync::RwLock;
 
 pub type Hash = <PoseidonHash as Hasher>::Hash;
 
-pub struct TreeState {
-    pub next_leaf:   usize,
-    pub merkle_tree: PoseidonTree,
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub struct TreeUpdate {
+    pub leaf_index: usize,
+    pub element:    Hash,
```

Comment:
If these are public they should have documentation.

Reply:
I have trouble understanding what needs documenting here.

Reply:
It's not clear to me what `leaf_index` actually entails. Is it the index at which the `element` is inserted?

Reply:
Yeah. A merkle tree holds elements in leaves only. And they are indexed. It's consistent with the usage in the database and in the contract too.